presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LimitPushDown.java

```diff
@@ -117,13 +125,43 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<LimitConte
 
-            if (limit != null &&
-                    node.getAggregations().isEmpty() &&
+            if (limit != null) {
+                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
+                if (node.getAggregations().isEmpty() &&
                     node.getOutputSymbols().size() == node.getGroupBy().size() &&
                     node.getOutputSymbols().containsAll(node.getGroupBy())) {
-                checkArgument(!node.getSampleWeight().isPresent(), "DISTINCT aggregation has sample weight symbol");
-                PlanNode rewrittenSource = context.rewrite(node.getSource());
-                return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), Optional.empty());
+                    PlanNode rewrittenSource = context.rewrite(node.getSource());
+                    return new DistinctLimitNode(idAllocator.getNextId(), rewrittenSource, limit.getCount(), Optional.empty(), node.getGroupBy(), Optional.empty(), Optional.empty());
+                }
+                else if (node.getOutputSymbols().size() == node.getGroupBy().size() + node.getAggregations().size() &&
+                        node.getFunctions().values().stream().allMatch(signature -> signature.getName().equals("arbitrary"))) {
+                    PlanNode rewrittenSource = context.rewrite(node.getSource());
+
+                    ImmutableMap.Builder<Symbol, Expression> aggregationMapBuilder = ImmutableMap.builder();
+                    for (Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                        aggregationMapBuilder.put(entry.getKey(), entry.getValue().getArguments().get(0));
+                    }
+                    Map<Symbol, Expression> aggregationSymbolMap = aggregationMapBuilder.build();
+
+                    List<Symbol> nonDistinctSymbols = aggregationSymbolMap.values().stream()
+                        .map(expression -> Symbol.fromQualifiedName(((QualifiedNameReference) expression).getName()))
+                        .collect(toList());
+
+                    PlanNode filterNode = rewrittenSource;
+                    for (Symbol nonDistinctSymbol : nonDistinctSymbols) {
+                        filterNode = new FilterNode(idAllocator.getNextId(), filterNode, new IsNotNullPredicate(nonDistinctSymbol.toQualifiedNameReference()));
+                    }
+
+                    ImmutableMap.Builder<Symbol, Expression> projectionMapBuilder = ImmutableMap.builder();
+                    projectionMapBuilder.putAll(aggregationSymbolMap);
+                    for (Symbol symbol : node.getGroupBy()) {
+                        projectionMapBuilder.put(symbol, symbol.toQualifiedNameReference());
+                    }
+                    Map<Symbol, Expression> projectionMap = projectionMapBuilder.build();
+
+                    PlanNode projectNode = new ProjectNode(idAllocator.getNextId(), filterNode, projectionMap);
```

Comment:
Might be missing something, but can you explain why we need this projection node?